azurelinuxagent/ga/firewall_manager.py

```diff
@@ -0,0 +1,499 @@
+        # has aged out, keep this cleanup in place.
+        return self._base_command + [command_option, "OUTPUT", "-d", self._wire_server_address, "-p", "tcp", "-m", "conntrack", "--ctstate", "INVALID,NEW", "-j", "ACCEPT"]
+
+    def _get_append_command_option(self):
+        return "-A"
+
+    def _get_check_command_option(self):
+        return "-C"
+
+    def _get_delete_command_option(self):
+        return "-D"
+
+
+class FirewallCmd(_FirewallManagerMultipleRules):
+    """
+    FirewallManager based on the firewalld command-line tool.
+    """
+    def _get_state_command(self):
+        return ["firewall-cmd", "--permanent", "--direct", "--get-all-passthroughs"]
+
+    def _get_accept_dns_rule_command(self, command_option):
+        return ["firewall-cmd", "--permanent", "--direct", command_option, "ipv4", "-t", "security", "-A", "OUTPUT", "-d", self._wire_server_address, "-p", "tcp", '--destination-port', '53', '-j', 'ACCEPT']
+
+    def _get_accept_rule_command(self, command_option):
+        return ["firewall-cmd", "--permanent", "--direct", command_option, "ipv4", "-t", "security", "-A", "OUTPUT", "-d", self._wire_server_address, "-p", "tcp", "-m", "owner", "--uid-owner", str(os.getuid()), "-j", "ACCEPT"]
+
+    def _get_drop_rule_command(self, command_option):
+        return ["firewall-cmd", "--permanent", "--direct", command_option, "ipv4", "-t", "security", "-A", "OUTPUT", "-d", self._wire_server_address, "-p", "tcp", "-m", "conntrack", "--ctstate", "INVALID,NEW", "-j", "DROP"]
+
+    def _get_legacy_rule_command(self, command_option):
+        # Agents <= 2.7.0.6 inserted (-I) the rule to accept DNS traffic; later agents changed that to append (-A) the rule.
+        # The insert rule needs to be removed, otherwise there will be duplicate rules for DNS.
+        return ["firewall-cmd", "--permanent", "--direct", command_option, "ipv4", "-t", "security", "-I", "OUTPUT", "-d", self._wire_server_address, "-p", "tcp", '--destination-port', '53', '-j', 'ACCEPT']
+
+    def _get_append_command_option(self):
+        return "--passthrough"
+
+    def _get_check_command_option(self):
+        return "--query-passthrough"
+
+    def _get_delete_command_option(self):
+        return "--remove-passthrough"
+
+
+class NfTables(FirewallManager):
+    """
+    FirewallManager based on the nft command-line tool.
+    """
+    def setup(self):
+        shellutil.run_command(["nft", "add", "table", "ip", "walinuxagent"])
+        shellutil.run_command(["nft", "add", "chain", "ip", "walinuxagent", "output", "{", "type", "filter", "hook", "output", "priority", "0", ";", "policy", "accept", ";", "}"])
+        shellutil.run_command([
+            "nft", "add", "rule", "ip", "walinuxagent", "output", "ip", "daddr", self._wire_server_address,
+            "tcp", " dport", "!=", "53",
+            "skuid", "!=", str(os.getuid()),
+            "ct", "state", "invalid,new", "counter", "drop"])
```

Comment:
I came across some interesting functionality for nfttables, I don't know it's worth in our code. Unlike iptables it provides atomicity when execute multiple rules at once.  

https://wiki.nftables.org/wiki-nftables/index.php/Atomic_rule_replacement

`With iptables it was common to use a bash script comprised of multiple iptables commands to configure a firewall. This is sub-optimal because it is not atomic, that is to say that during the few fractions of a second that your bash script takes to run your firewall is in a partially configured state. Nftables introduces atomic rule replacement with the -f option. This is different from bash scripts because nftables will read in all of the included config files, create the config object in memory alongside the existing config, and then in one atomic operation it swaps the old config for the new one meaning there is no moment when the firewall is partially configured`

Reply:
Yes, it is something we should take advantage of. Same thing with firewalld, we are not taking advantage of all its functionality. I posted this in our internal chat when I posted my PR. Will be a new iteration in this code.